test/decl/class/circular_inheritance.swift

```diff
@@ -72,3 +72,12 @@ func crash() {
   Circle()
   // expected-error@-1 {{'Circle' cannot be constructed because it has no accessible initializers}}
 }
+
+// FIXME: We shouldn't emit the redundant "circular reference" diagnostics here.
+class WithDesignatedInit : WithDesignatedInit {
+  // expected-error@-1 {{'WithDesignatedInit' inherits from itself}}
+  // expected-error@-2 {{circular reference}}
+  // expected-note@-3 {{through reference here}}
+
+  init(x: Int) {} // expected-error {{circular reference}}
```

Comment:
If you wanted to emit a better diagnostic, you could override `diagnoseCycle` and try to detect this case.  It's probably not worth the extra hassle in this patch.

See also rdar://55263708

Reply:
Yeah, I talked with Slava about this – the goal is to eventually have a principled way to avoid these kinds of duplicate cycle diagnostics. For now though, I think it's okay.